### PR TITLE
Signin

### PR DESCRIPTION
**Pull request by gonziesc**
travis
## Summary
- Sign in implemented
- Post http request from “/users/sessions”  done
- Returns authorization token
- Tests
- Added errors 
- Added min length property to password

## Screenshots
screenshot from 2017-04-19 11-12-25
screenshot from 2017-04-19 11-12-37
screenshot from 2017-04-19 11-12-43

## Trello Cards
https://trello.com/c/n0MjCqmg/3-sign-in
https://trello.com/c/nSdgOh8K/2-sign-up

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -1,30 +1,50 @@
...
-  const errorsSend = userService.validateUser(name, surname, mail, passw);
-  if (passw.length <= 8) {
+  const user = req.body ? {
+    firstName: req.body.firstName,
+    lastName: req.body.lastName,
+    password: req.body.password,
+    email: req.body.email
+  } : {};
+  const errorsSend = userService.validateUser(user.firstName, user.lastName, user.email);
+  if (user.password.length <= 8) {
```
Comment (mishuagopian) [inline app/controllers/users.js]: This should be `<`

File: app/controllers/users.js [hunk 2]
```diff
@@ -1,30 +1,50 @@
...
     }).catch((err) => {
       next(errors.defaultError(err));
     });
+  }
+};
+exports.login = (req, res, next) => {
+  const password = req.body.password;
+  const email = req.body.email;
+  const errorsSend = userService.validateEmail(email);
+  if (errorsSend.length === 0) {
```
Comment (mishuagopian) [inline app/controllers/users.js]: I think doing `if (errorsSend.length) { ... }` case first is cleaner.

File: app/controllers/users.js [hunk 3]
```diff
@@ -1,30 +1,50 @@
...
+    userService.getByMail(email).then((u) => {
+      if (u) {
+        if (password === u.password) {
+          const token = session.generateAccessToken(u);
+          console.log(token);
```
Comment (mishuagopian) [inline app/controllers/users.js]: Remove console log

File: app/errors.js [hunk 4]
```diff
@@ -4,6 +4,18 @@ exports.notFound = {
   message: 'Not found'
 };
 
+exports.invalidUser = {
+  statusCode: 402,
```
Comment (mishuagopian) [inline app/errors.js]: Just 400: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_Client_errors

File: app/errors.js [hunk 5]
```diff
@@ -4,6 +4,18 @@ exports.notFound = {
...
+  message: 'Invalid username or password'
+};
+
+exports.invalidCreate = function (messageReturned) {
```
Comment (mishuagopian) [inline app/errors.js]: Remember using es6 features:
```node
exports.invalidCreate = (message) => ({
  statusCode:400,
  message
});
```

File: app/controllers/users.js [hunk 6]
```diff
@@ -1,30 +1,50 @@
...
+          res.status(200);
+          res.send(u);
```
Comment (mishuagopian) [inline app/controllers/users.js]: Token should be returned and not the user.

File: app/services/sessionManager.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  config = require('./../../config/config').config;
+
+const SECRET = config.common.session.secret;
+
+exports.generateAccessToken = (user) => {
+  const claims = { sub: user.firstName,
+    iss: '/users/sessions',
+    expires_in: '3600',
+    permissions: 'access'
+  };
```
Comment (mishuagopian) [inline app/services/sessionManager.js]: We should enode data which may help us relate the token with one user, so a unique value must be included in the object to encode. Maybe we can also add a an expire date and not some time as you did..

File: app/services/users.js [hunk 8]
```diff
@@ -31,3 +30,31 @@ exports.validateUser = function (name, lastName, mail) {
   }
   return errorsReturn;
 };
+
+exports.validateEmail = function (mail) {
```
Comment (mishuagopian) [inline app/services/users.js]: All this code is duplicated.
Also, what about the regex I gave you [here](https://github.com/wolox-training/ge-express-js/pull/3#discussion_r112185231)

File: app/services/users.js [hunk 9]
```diff
@@ -31,3 +30,31 @@ exports.validateUser = function (name, lastName, mail) {
...
+  }
+};
+
+exports.getOne = (user) => {
+  return orm.models.user.findOne({ where: user }).catch((err) => {
+    throw errors.databaseError(err.detail);
+  });
+};
+
+exports.getByMail = (email) => {
```
Comment (mishuagopian) [inline app/services/users.js]: nice handled!

File: test/users.js [hunk 10]
```diff
@@ -9,9 +9,9 @@ describe('/users POST', () => {
     .post('/users')
     .send({ firstName: 'firstName', lastName: 'lastName', password: 'password888' })
     .catch((err) => {
-      err.should.have.status(401);
-      const errorsRecieved = err.response.text;
-      errorsRecieved.should.be.equal('["The mail field can not be empty"]');
+      err.should.have.status(400);
+      const errorsRecieved = err.response.body.error;
+      assert.equal('The mail field can not be empty', errorsRecieved);
```
Comment (mishuagopian) [inline test/users.js]: You should use the `should` syntax as you were doing before.

File: test/users.js [hunk 11]
```diff
@@ -74,9 +74,54 @@ describe('/users POST', () => {
...
-        err.should.have.status(401);
-        const errorsRecieved = err.response.text;
-        errorsRecieved.should.be.equal('["The password has to have 8 or more characters"]');
+        err.should.have.status(400);
+        const errorsRecieved = err.response.body.error;
+        assert.equal('The password has to have 8 or more characters', errorsRecieved[0]);
       }).then(() => done());
   });
+
+  it('should be successful bringing the user', (done) => {
```
Comment (mishuagopian) [inline test/users.js]: Should return the access token

File: test/users.js [hunk 12]
```diff
@@ -74,9 +74,54 @@ describe('/users POST', () => {
...
+    chai.request(server)
+      .post('/users/sessions')
+      .send({ password: 'usernaasdasdme3', email: 'email1@wolox.com.ar' })
+      .then((res) => {
+        res.should.have.status(200);
```
Comment (mishuagopian) [inline test/users.js]: Verify access token is returned

File: test/users.js [hunk 13]
```diff
@@ -74,9 +74,54 @@ describe('/users POST', () => {
...
+      }).then(() => done());
+  });
+
+  it('should fail because mail is invalid', (done) => {
+    chai.request(server)
+      .post('/users/sessions')
+      .send({ password: 'usernaasdasdme3', email: 'email1wolox.com.ar' })
+      .catch((err) => {
+        err.should.have.status(400);
+        const errorsRecieved = err.response.body;
```
Comment (mishuagopian) [inline test/users.js]: Why are you sometimes using `err.response.error` and sometimes `err.response.body`?

File: test/users.js (hunk 11 above)
Comment (mishuagopian) [inline test/users.js]: All these `it`s should be in a different `describe` as you are testing a new endpoint

**Comment by gonziesc** [comment]
travis
## Summary
- Verified access token return
- Used only err.response.body

## Trello Cards
https://trello.com/c/n0MjCqmg/3-sign-in

File: app/controllers/users.js [hunk 14]
```diff
@@ -18,5 +20,31 @@ exports.create = (req, res, next) => {
...
       next(errors.defaultError(err));
     });
+
+  }
+};
+exports.login = (req, res, next) => {
+  const password = req.body.password;
+  const email = req.body.email;
+  const errorsSend = [];
+  userService.validateEmail(email, errorsSend);
```
Comment (mishuagopian) [inline app/controllers/users.js]: No need to send the array, you can do:
```node
const errorsSend = userService.validateEmail(email);
```

File: app/controllers/users.js [hunk 15]
```diff
@@ -18,5 +20,31 @@ exports.create = (req, res, next) => {
...
+  if (errorsSend.length) {
+    res.status(400);
+    res.send(errorsSend);
+    res.end();
```
Comment (mishuagopian) [inline app/controllers/users.js]: You can remove this as sending a message closes the stream

File: app/controllers/users.js [hunk 16]
```diff
@@ -18,5 +20,31 @@ exports.create = (req, res, next) => {
...
+  } else {
+    userService.getByMail(email).then((u) => {
+      if (u) {
+        if (password === u.password) {
+          const accessToken = session.generateAccessToken(u.email);
+          res.status(200);
+          res.send(accessToken);
+        } else {
+          next(errors.invalidPassword);
```
Comment (mishuagopian) [inline app/controllers/users.js]: I'd always send `errors.invalidUser` without specifying if the wront parameter was the email or the password.

File: app/errors.js [hunk 17]
```diff
@@ -4,6 +4,21 @@ exports.notFound = {
   message: 'Not found'
 };
 
+exports.invalidUser = {
+  statusCode: 400,
+  message: 'Invalid username'
```
Comment (mishuagopian) [inline app/errors.js]: Invalid email or passowrd

File: app/services/users.js [hunk 18]
```diff
@@ -7,27 +7,45 @@ exports.create = (user) => {
...
-  if (!lastName) {
+  if (!user.lastName) {
     errorsReturn.push('The last name field can not be empty');
   }
-  if (mail) {
-    const domain = mail.substring(mail.lastIndexOf('@') + 1);
-    /* eslint max-len: off */
-    const re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
-    if (!re.test(mail)) {
+  exports.validateEmail(user.email, errorsReturn);
```
Comment (mishuagopian) [inline app/services/users.js]: Don;t send the errors array:
```node
const emailError = exports.evalidateEmail(user.email);
if (emailError) {
  errorsReturn.append(emailError);
}

File: app/services/users.js [hunk 19]
```diff
@@ -7,27 +7,45 @@ exports.create = (user) => {
...
+  return errorsReturn;
+};
+
+exports.validateEmail = (email, errorsReturn) => {
```
Comment (mishuagopian) [inline app/services/users.js]: Don't receive errors array, just define one in this function.

File: test/users.js [hunk 20]
```diff
@@ -74,8 +72,55 @@ describe('/users POST', () => {
...
+      }).then(() => done());
+  });
+
+  it('should be successful bringing the token', (done) => {
+    chai.request(server)
+      .post('/users/sessions')
+      .send({ password: 'usernaasdasdme3', email: 'email1@wolox.com.ar' })
+      .then((res) => {
+        res.should.have.status(200);
+        res.text.length.should.be.above(20);
```
Comment (mishuagopian) [inline test/users.js]: Check access token is received and not this.

File: test/users.js (hunk 13 above)
Comment (mishuagopian) [inline test/users.js]: answer this please

File: test/users.js [hunk 21]
```diff
@@ -74,8 +72,55 @@ describe('/users POST', () => {
...
   });
+
+  it('should fail because mail is invalid', (done) => {
+    chai.request(server)
+      .post('/users/sessions')
+      .send({ password: 'usernaasdasdme3', email: 'email1wolox.com.ar' })
+      .catch((err) => {
+        err.should.have.status(400);
+        const errorsRecieved = err.response.body;
+        assert.equal('The mail is not valid', errorsRecieved[0]);
```
Comment (mishuagopian) [inline test/users.js]: Use should syntax

File: test/users.js [hunk 22]
```diff
@@ -74,8 +72,55 @@ describe('/users POST', () => {
...
+
+  it('should fail because password is incorrect', (done) => {
+    chai.request(server)
+      .post('/users/sessions')
+      .send({ password: 'usernaasdasdme3', email: 'email1@wolox.com.ar' })
+      .catch((err) => {
+        err.should.have.status(400);
+        err.response.should.be.json;
+        const errorsRecieved = err.response.body.error;
+        assert.equal('Invalid password', errorsRecieved);
```
Comment (mishuagopian) [inline test/users.js]: should syntax

**Pull request merged**